File: iOS-trip-search/AddressSettingViewController.m
//
//  AddressSettingViewController.m
//  iOS-trip-search
//
//  Created by hanxiaoming on 2017/5/27.
//  Copyright © 2017年 Amap. All rights reserved.
//

#import "AddressSettingViewController.h"
#import <AMapSearchKit/AMapSearchKit.h>

#import "MyCityListView.h"
#import "MySearchResultView.h"
#import "MySearchBarView.h"

#import "MyCityManager.h"

#define kTableViewMargin    8
#define kNaviBarHeight      60

@interface AddressSettingViewController ()<AMapSearchDelegate, MyCityListViewDelegate, MySearchBarViewDelegate, MySearchResultViewDelegate>

@property (nonatomic, strong) AMapSearchAPI *search;

@property (nonatomic, strong) UIButton *titleButton;

@property (nonatomic, strong) UIView *listContainerView;

@property (nonatomic, strong) MyCityListView *cityListView;
@property (nonatomic, strong) MySearchResultView *searchResultView;

@property (nonatomic, strong) MySearchBarView *searchBar;

@property (nonatomic, strong) AMapPOIKeywordsSearchRequest *currentRequest;

@end

@implementation AddressSettingViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    self.navigationController.navigationBar.translucent = NO;
    self.extendedLayoutIncludesOpaqueBars = YES;
    
    [self.navigationController setNavigationBarHidden:YES];
    
    self.view.backgroundColor = [UIColor lightGrayColor];
    
    if (_searchTextPlaceholder == nil) {
        _searchTextPlaceholder = @"出入搜索关键字";
    }
    [self initSearch];
    [self initSearchBarView];
    [self initListContainerView];
    
    //update
    [self updateCurrentCity:self.currentCity force:YES];
}

- (void)viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];
    [self.searchBar becomeFirstResponder];
}

- (void)viewWillDisappear:(BOOL)animated
{
    [super viewWillDisappear:animated];
    [self.searchBar becomeFirstResponder];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - initialization

- (void)initSearch
{
    //search
    self.search = [[AMapSearchAPI alloc] init];
    self.search.delegate = self;
}

- (void)initSearchBarView
{
    self.searchBar = [[MySearchBarView alloc] initWithFrame:CGRectMake(0, 0, self.view.bounds.size.width, kNaviBarHeight)];
    self.searchBar.doubleSearchModeEnable = YES;
    self.searchBar.delegate = self;
    self.searchBar.searchTextPlaceholder = _searchTextPlaceholder;
    
    [self.view addSubview:self.searchBar];
}

- (void)initListContainerView
{
    self.listContainerView = [[UIView alloc] initWithFrame:CGRectMake(kTableViewMargin, kTableViewMargin + kNaviBarHeight, self.view.bounds.size.width - kTableViewMargin * 2, self.view.bounds.size.height - kTableViewMargin - kNaviBarHeight)];
    self.listContainerView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    self.listContainerView.layer.shadowOpacity = 0.3;
    self.listContainerView.layer.shadowOffset = CGSizeMake(0, 0.5);
    [self.view addSubview:self.listContainerView];
    
    
    //
    _cityListView = [[MyCityListView alloc] initWithFrame:self.listContainerView.bounds];
    _cityListView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    _cityListView.delegate = self;
    
    [self.listContainerView addSubview:_cityListView];
    
    _searchResultView = [[MySearchResultView alloc] initWithFrame:self.listContainerView.bounds];
    _searchResultView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    _searchResultView.delegate = self;
    
    //not show the address setting cell
    _searchResultView.showsAddressSettingCell = NO;
    
    [self.listContainerView addSubview:_searchResultView];
}

#pragma mark - handler

- (void)updateCurrentCity:(MyCity *)currentCity force:(BOOL)force
{
    MyCity *oldCity = self.currentCity;
    
    _currentCity = currentCity;
    
    if (self.search) {
        self.searchBar.seachCity = currentCity;
        
        // 城市改变后，或者需要强制搜索
        if (force || ![oldCity.name isEqualToString:_currentCity.name]) {
            [self searchPoiByKeyword:self.searchBar.currentSearchKeywords city:self.currentCity];
        }
    }
}

- (void)searchPoiByKeyword:(NSString *)keyword city:(MyCity *)city
{
    AMapPOIKeywordsSearchRequest *request = [[AMapPOIKeywordsSearchRequest alloc] init];
    request.keywords = keyword;
    request.cityLimit = YES;
    
    request.city = city.name;
    
    //TODO: 需要设置location和sortrule
    
    [self.search AMapPOIKeywordsSearch:request];
    
    self.currentRequest = request;
}

#pragma mark - MySearchBarViewDelegate

- (void)searchBarView:(MySearchBarView *)searchBarView didSearchTextChanged:(NSString *)text
{
    if (!self.searchBar.doubleSearchModeEnable) {
        self.cityListView.filterKeywords = text;
    }
    else {
        [self searchPoiByKeyword:text city:self.currentCity];
    }
}

- (void)searchBarView:(MySearchBarView *)searchBarView didCityTextChanged:(NSString *)text
{
    if (self.searchBar.doubleSearchModeEnable) {
        self.cityListView.filterKeywords = text;
    }
}

- (void)didCancelButtonTapped:(MySearchBarView *)searchBarView
{
    if ([self.delegate respondsToSelector:@selector(didCancelButtonTappedForAddressSettingViewController:)]) {
        [self.delegate didCancelButtonTappedForAddressSettingViewController:self];
    }
}

- (void)searchBarView:(MySearchBarView *)searchBarView didCityTextShown:(BOOL)shown
{
    self.searchResultView.hidden = shown;
}

#pragma mark - MyCityListViewDelegate

- (void)cityListView:(MyCityListView *)listView didCitySelected:(MyCity *)city
{
    [self updateCurrentCity:city force:NO];
}

- (void)didCityListViewwScroll:(MyCityListView *)listView
{
    [self.searchBar resignFirstResponder];
}


#pragma mark - MySearchResultViewDelegate

- (void)resultListView:(MySearchResultView *)listView didPOISelected:(AMapPOI *)poi
{
    if ([self.delegate respondsToSelector:@selector(addressSettingViewController:didPOISelected:)]) {
        [self.delegate addressSettingViewController:self didPOISelected:poi];
    }
}

- (void)didResultListViewScroll:(MySearchResultView *)listView
{
    [self.searchBar resignFirstResponder];
}

#pragma mark - AMapSearch

- (void)AMapSearchRequest:(id)request didFailWithError:(NSError *)error
{
    NSLog(@"search error :%@", error);
}

- (void)onPOISearchDone:(AMapPOISearchBaseRequest *)request response:(AMapPOISearchResponse *)response
{
    if (self.currentRequest == request) {
        self.searchResultView.poiArray = response.pois;
    }
}

//- (void)onReGeocodeSearchDone:(AMapReGeocodeSearchRequest *)request response:(AMapReGeocodeSearchResponse *)response
//{
//    if (response.regeocode == nil) {
//        return;
//    }
//    
//    if ([MyCityManager sharedInstance].locationCity == nil) {
//        [MyCityManager sharedInstance].locationCity = [[MyCity alloc] init];
//        [MyCityManager sharedInstance].locationCity.name = response.regeocode.addressComponent.city;
//        
//        self.cityListView.locationCity = [MyCityManager sharedInstance].locationCity;
//        
//        if ([MyCityManager sharedInstance].currentCity == nil) {
//            [self updateCurrentCity:[MyCityManager sharedInstance].locationCity];
//        }
//        
//    }
//    
//    // just regeo for poi
//    self.locationView.startPOI = response.regeocode.pois.firstObject;
//    [self addPositionAnnotation:self.startAnnotation forPOI:self.locationView.startPOI];
//}

//- (void)onGeocodeSearchDone:(AMapGeocodeSearchRequest *)request response:(AMapGeocodeSearchResponse *)response
//{
//    if (response.geocodes.count == 0)
//    {
//        return;
//    }
//    AMapGeocode *geocode = response.geocodes.firstObject;
//    [self searchReGeocodeWithLocation:geocode.location];
//    
//    NSLog(@"move to %@ %@", geocode.city, geocode.location.formattedDescription);
//}

@end

File: iOS-trip-search-swift/AddressSettingViewController.swift
//
//  AddressSettingViewController.swift
//  iOS-trip-search
//
//  Created by hanxiaoming on 2017/6/5.
//  Copyright © 2017年 Amap. All rights reserved.
//

import UIKit

fileprivate let kTableViewMargin: CGFloat = 8.0
fileprivate let kNaviBarHeight: CGFloat = 60.0

protocol AddressSettingViewControllerDelegate: NSObjectProtocol {
    
    func addressSettingViewController(_ viewController: AddressSettingViewController, didPOISelected poi: MyLocation!)

    func didCancelButtonTappedForAddressSettingViewController(_ viewController: AddressSettingViewController)
}

class AddressSettingViewController: UIViewController, AMapSearchDelegate, MyCityListViewDelegate, MySearchBarViewDelegate, MySearchResultViewDelegate {

    private var search: AMapSearchAPI!
    private var titleButton: UIButton!
    
    private var listContainerView: UIView!
    private var cityListView: MyCityListView!
    private var searchBar: MySearchBarView!
    private var searchResultView: MySearchResultView!
    
    private var currentRequest: AMapInputTipsSearchRequest?
    
    
    public weak var delegate: AddressSettingViewControllerDelegate?
    public var currentCity: MyCity?
    public var searchTextPlaceholder: String?
    
    override func viewDidLoad() {
        super.viewDidLoad()

        // Do any additional setup after loading the view.
        self.navigationController?.navigationBar.isTranslucent = false
        self.extendedLayoutIncludesOpaqueBars = true
        
        self.navigationController?.setNavigationBarHidden(true, animated: false)
        
        self.view.backgroundColor = UIColor.lightGray
        
        if (searchTextPlaceholder == nil) {
            searchTextPlaceholder = "请输入搜索关键字"
        }
        initSearch()
        initSearchBarView()
        initListContainerView()
        
        //update
        updateCurrentCity(self.currentCity, forceSearchPOI: true)
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        searchBar.becomeFirstResponder()
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        searchBar.resignFirstResponder()
    }

    //MARK: - init
    func initSearch() {
        self.search = AMapSearchAPI()
        self.search.delegate = self
    }
    
    func initSearchBarView() {
        self.searchBar = MySearchBarView(frame: CGRect(x: 0.0, y: 0.0, width: self.view.bounds.width, height: kNaviBarHeight))
        self.searchBar.doubleSearchModeEnable = true
        self.searchBar.delegate = self
        self.searchBar.searchTextPlaceholder = self.searchTextPlaceholder
        self.view.addSubview(self.searchBar)
    }
    
    func initListContainerView() {
        self.listContainerView = UIView(frame: CGRect(x: kTableViewMargin, y: kTableViewMargin + kNaviBarHeight, width: self.view.bounds.width - kTableViewMargin * 2, height: self.view.bounds.height - kTableViewMargin - kNaviBarHeight))
        
        
        self.listContainerView.autoresizingMask = [UIViewAutoresizing.flexibleHeight, UIViewAutoresizing.flexibleWidth]
        
        self.listContainerView.layer.shadowOpacity = 0.3
        self.listContainerView.layer.shadowOffset = CGSize(width: 0, height: 0.5)
        self.view.addSubview(self.listContainerView)
        
        //
        self.cityListView = MyCityListView(frame: self.listContainerView.bounds)
        self.cityListView.autoresizingMask = [UIViewAutoresizing.flexibleHeight, UIViewAutoresizing.flexibleWidth]
        self.cityListView.delegate = self
        
        self.listContainerView.addSubview(self.cityListView)
        
        //
        self.searchResultView = MySearchResultView(frame: self.listContainerView.bounds)
        self.searchResultView.autoresizingMask = [UIViewAutoresizing.flexibleHeight, UIViewAutoresizing.flexibleWidth]
        self.searchResultView.delegate = self
        self.searchResultView.showsAddressSettingCell = false
        self.listContainerView.addSubview(self.searchResultView)
    }

    //MARK: - handler
    func updateCurrentCity(_ currentCity: MyCity!, forceSearchPOI force: Bool) {
        let oldCity = self.currentCity
        
        self.currentCity = currentCity
        self.searchBar.seachCity = currentCity
        
        // 城市改变后，或者需要强制搜索
        if force || !(oldCity?.name == self.currentCity?.name) {
            self.searchTipsByKeyword(self.searchBar.currentSearchKeywords, inCity: self.currentCity)
        }
        
    }
    
//    func searchPoiByKeyword(_ keyword: String?, inCity city: MyCity?) {
//        
//        let request = MyRecordManager.poiSearchRequest(withKeyword: keyword, in: city)
//        search.aMapPOIKeywordsSearch(request)
//        currentRequest = request
//    }
    
    func searchTipsByKeyword(_ keyword: String?, inCity city: MyCity?) {
        if keyword?.characters.count == 0 {
            return
        }
        
        let request = AMapInputTipsSearchRequest()
        
        request.city = city?.name
        request.cityLimit = true
        request.keywords = keyword
        search.aMapInputTipsSearch(request)
        currentRequest = request
    }
    
    //MARK: - MySearchBarViewDelegate
    func searchBarView(_ searchBarView: MySearchBarView!, didSearchTextChanged text: String!) {
        if !self.searchBar.doubleSearchModeEnable {
            self.cityListView.filterKeywords = text
        }
        else {
            self.searchTipsByKeyword(text, inCity: self.currentCity)
        }
    }
    
    func searchBarView(_ searchBarView: MySearchBarView!, didCityTextChanged text: String!) {
        if self.searchBar.doubleSearchModeEnable {
            self.cityListView.filterKeywords = text
        }
    }
    
    func searchBarView(_ searchBarView: MySearchBarView!, didCityTextShown shown: Bool) {
        self.searchResultView.isHidden = shown
    }
    
    func didCancelButtonTapped(_ searchBarView: MySearchBarView!) {
        if delegate != nil {
            delegate?.didCancelButtonTappedForAddressSettingViewController(self)
        }
    }
    
    //MARK: - MyCityListViewDelegate
    
    func cityListView(_ listView: MyCityListView!, didCitySelected city: MyCity!) {
        self.updateCurrentCity(city, forceSearchPOI: false)
    }
    
    func didCityListViewwScroll(_ listView: MyCityListView!) {
        self.searchBar.resignFirstResponder()
    }
    
    //MARK: - MySearchResultViewDelegate
    
    
    func resultListView(_ listView: MySearchResultView!, didPOISelected poi: MyLocation!) {
        if delegate != nil {
            delegate?.addressSettingViewController(self, didPOISelected: poi)
        }
    }
    
    func didResultListViewScroll(_ listView: MySearchResultView!) {
        self.searchBar.resignFirstResponder()
    }
    
    //MARK: - AMapSearch
    func aMapSearchRequest(_ request: Any!, didFailWithError error: Error!) {
        print("search error :\(error)")
    }
    
    func onInputTipsSearchDone(_ request: AMapInputTipsSearchRequest!, response: AMapInputTipsSearchResponse!) {
        if self.currentRequest != nil && self.currentRequest! == request {
            
            var locations = [MyLocation]()
            for tip in response.tips {
                let loc = MyLocation(tip: tip, city: request.city)
                
                if loc != nil {
                    locations.append(loc!)
                }
            }
            self.searchResultView.poiArray = locations;
        }

    }
    
//    func onPOISearchDone(_ request: AMapPOISearchBaseRequest!, response: AMapPOISearchResponse!) {
//        if self.currentRequest != nil && self.currentRequest! == request {
//            self.searchResultView.poiArray = response.pois
//        }
//    }
//    
}
